Yosemite/Yosemite/Stores/MetaDataStore.swift

```diff
@@ -0,0 +1,209 @@
+        newStorageMetaData.update(with: readOnlyMetaData)
+        newStorageMetaData.order = storageOrder
+        storageOrder.addToCustomFields(newStorageMetaData)
+    }
+
+    /// Updates product metadata in the local database.
+    /// - Parameters:
+    ///   - siteID: Site id of the product.
+    ///   - productID: ID of the product.
+    ///   - metadata: Metadata to be updated.
+    ///   - onCompletion: Callback called when the action is finished, including the result of the update operation.
+    ///
+    func updateProductMetaData(siteID: Int64,
+                               productID: Int64,
+                               metadata: [[String: Any]],
+                               onCompletion: @escaping (Result<[MetaData], Error>) -> Void) {
+        Task { @MainActor in
+            do {
+                let results = try await self.remote.updateMetaData(for: siteID, for: productID, type: .product, metadata: metadata)
+                upsertStoredProductMetaDataInBackground(readOnlyProductMetaDatas: results, productID: productID, siteID: siteID) {
+                    onCompletion(.success(results))
+                }
+            } catch {
+                onCompletion(.failure(error))
+            }
+        }
+    }
+
+    /// Updates (OR Inserts) the specified ReadOnly ProductNote Entities into the Storage Layer.
+    /// - Parameters:
+    ///   - readOnlyProductMetaDatas: Array of read-only product metadata.
+    ///   - productID: ID of the product.
+    ///   - siteID: Site id of the product.
+    ///   - onCompletion: Callback called when the action is finished.
+    ///
+    func upsertStoredProductMetaDataInBackground(readOnlyProductMetaDatas: [Networking.MetaData], 
+                                                 productID: Int64,
+                                                 siteID: Int64,
+                                                 onCompletion: @escaping () -> Void) {
+        let derivedStorage = sharedDerivedStorage
+        derivedStorage.perform {
+            for readOnlyProductMetaData in readOnlyProductMetaDatas {
+                self.saveMetaData(derivedStorage, readOnlyProductMetaData, productID: productID, siteID: siteID)
+            }
+        }
+
+        storageManager.saveDerivedType(derivedStorage: derivedStorage) {
+            DispatchQueue.main.async(execute: onCompletion)
+        }
+    }
+
+    /// Using the provided StorageType, update or insert a Storage.MetaData using the provided ReadOnly
+    /// MetaData. This func does *not* persist any unsaved changes to storage.
+    /// - Parameters:
+    ///   - storage: The storage type to use.
+    ///   - readOnlyMetaData: The read-only metadata to save.
+    ///   - productID: ID of the product.
+    ///   - siteID: Site id of the product.
+    ///
+    private func saveMetaData(_ storage: StorageType, _ readOnlyMetaData: MetaData, productID: Int64, siteID: Int64) {
```

Comment:
The `private` is not needed as this is already inside a private extension.

Reply:
👍 a7e0b1ffb4494fc1f30eaa5431493ece830bb79a